proposals/0248-tpu-feedback.md

```diff
@@ -0,0 +1,186 @@
+---
+simd: '0248'
+title: TPU Feedback
+authors:
+  - Lijun Wang <lijun@anza.xyz>
+category: Standard
+type: Core
+status: Review
+created: 2025-02-14
+development:
+  - Anza - WIP
+  - Firedancer - Not started
+---
+
+## Summary
+
+This document defines the protocol to send feedback on TPU processing from the
+servers to the clients.
+
+## Motivation
+
+A client sends TPU transactions to the leader as QUIC uni-streams. Currently we
+only send transport level acknowledememt to the clients. I.e we only tell if a
+transaction has been received by the server. A transaction received by the
+server may not be processed due to resource contention. For example, if there
+are too many transactions with higher paying fees than this transaction, the
+transaction is not included in the block. Providing further information on the
+transaction state will help the client to make more informed decision on
+handling transaction failures and reduce the incentive to blindly resubmitting
+transactions and help reduce congestion condition in the network. The QUIC
+client library can be enhanced to use this information to not send transactions
+with fees too low.
+
+In addition, on the server side, the TPU server could use the feedback
+information to more accurately perform QOS. For example, if a peer is
+persistently submitting high percentage pf transactions which are rejected,
+its permitted throughput might be reduced.
+
+## Alternatives Considered
+
+There are different transport mechanism to send the feedback such as
+bidirectional streams, reverse unistream from the server to the client and
+QUIC datagrams. We choose datagrams due to its simplicity. One reason we have
+used stream over datagrams is the expectation that transaction size can be
+increased in the future which can not be carried in one datagram. Sending
+feedback has no such concern as we can send multiple datagrams if there are
+too many transactions to acknowledge.
+
+## New Terminology
+
+In this document we define the following,
+
+TPU Feedback -- information sent by the server about transaction processing.
+Server -- the validator receiving the TPU transactions and sending the feedback.
+Client -- the validator or application using QUIC TPU client sending the TPU
+          transactions and receiving the feedback.
+
+## Detailed Design
+
+When the client sends transactions to the server, it must first open a QUIC
+connection to the server. The feedback is sent by the server using the same
+connection. The feedback is sent using a QUIC datagram. The client wishing to
+receive feedback from the server must be ready to receive application datagrams
+from its connections.
+
+A client connection notifies the server of its interest to receive feedback by
+sending a datagram soliciting feedback. This helps server to avoid sending
+datagrams to clients which are not interested in receiving feedback or clients
+which are not upgraded to have such capabilities.
```

Comment:
I think it is best to assume we ask for feedback in the transaction request itself. Sending the feedback request in a separate datagram is adding statful behavior where it is not needed.

Reply:
There is a problem with that -- the transaction format need to be modified  and changes that protocol-- I think we run out of free flags.  I think this is a way to show the interest with minimal performance and protocol impact.

Reply:
do we not have some legacy variants we could reuse?